proto/control-api/src/channel/client.rs

```diff
@@ -0,0 +1,175 @@
+//! [`ControlApi`] client and [`CallbackApi`] server [`channel`]
+//! implementations.
+//!
+//! [`channel`]: futures::channel
+
+use async_trait::async_trait;
+use derive_more::{Display, Error, From};
+use futures::{
+    channel::{mpsc, oneshot},
+    StreamExt as _,
+};
+
+use crate::{
+    control, member, CallbackApi, ControlApi, Elements, Fid, Ping, Pong,
+};
+
+use super::{CallbackApiRequest, ControlApiRequest, SendErr};
+
+/// [`channel`]-based [`CallbackApi`] server.
+///
+/// [`channel`]: futures::channel
+#[derive(Debug)]
+pub struct CallbackApiServer<T: CallbackApi> {
+    /// Inner [`CallbackApi`] implementation.
+    pub(crate) api: T,
+
+    /// [`mpsc::UnboundedReceiver`] to receive [`CallbackApiRequest`].
+    pub(crate) receiver: mpsc::UnboundedReceiver<CallbackApiRequest<T::Error>>,
+}
+
+impl<T: CallbackApi> CallbackApiServer<T> {
+    /// Runs this [`CallbackApiServer`]. Completes after all
+    /// [`CallbackApiClient`]s linked to this [`CallbackApiServer`] are dropped.
+    ///
+    /// # Errors
+    ///
+    /// In case failed to send response via [`oneshot::Sender`].
+    ///
+    /// [`CallbackApiClient`]: super::CallbackApiClient
+    #[allow(clippy::map_err_ignore)]
+    pub async fn run(mut self) -> Result<(), SendErr> {
+        while let Some(ev) = self.receiver.next().await {
+            ev.sender
+                .send(self.api.on_event(ev.request).await)
+                .map_err(|_| SendErr)?;
```

Comment:
@ilslv well, that's not very wise implementation as executes requests strictly one-by-one (starts executing next one only after previous one finishes). We should execute them concurrently, optionally bounding the number of concurrently executing requests. I think `TryStreamExt::try_for_each()`/`TryStreamExt::try_for_each_concurrent()` will do here.

Also, finishing the whole server on a response `SendErr` seems also wrong. Does our gRPC implementation do so? 🙃